backend/src/main/java/ai/verta/modeldb/versioning/blob/DatasetContainer.java

```
@@ -0,0 +1,127 @@
+      throws NoSuchAlgorithmException {
+    final List<String> locationList = getLocationList();
+    String blobType = getBlobType();
+
+    TreeElem treeChildDataset =
+        rootTree.push(
+            locationList,
+            null,
+            blobType,
+            null); // need to ensure dataset is sorted
+    switch (dataset.getContentCase()) {
+      case S3:
+        for (S3DatasetComponentBlob componentBlob : dataset.getS3().getComponentsList()) {
+          final String sha256 = computeSHA(componentBlob);
+          S3DatasetComponentBlobEntity s3DatasetComponentBlobEntity =
+              new S3DatasetComponentBlobEntity(sha256, componentBlob);
+          treeChildDataset.push(
+              Arrays.asList(
+                  locationList.get(locationList.size() - 1), componentBlob.getPath().getPath()),
+              computeSHA(componentBlob.getPath()),
+              componentBlob.getClass().getSimpleName(),
+              s3DatasetComponentBlobEntity);
+        }
+        break;
+      case PATH:
+        for (PathDatasetComponentBlob componentBlob : dataset.getPath().getComponentsList()) {
+          final String sha256 = computeSHA(componentBlob);
+          PathDatasetComponentBlobEntity pathDatasetComponentBlobEntity =
+              new PathDatasetComponentBlobEntity(sha256, componentBlob);
+          treeChildDataset.push(
+              Arrays.asList(locationList.get(locationList.size() - 1), componentBlob.getPath()),
+              computeSHA(componentBlob),
+              componentBlob.getClass().getSimpleName(),
+              pathDatasetComponentBlobEntity);
+        }
+        break;
+      default:
+        break;
+    }
+  }
+
+  private String getBlobType() {
+    switch (dataset.getContentCase()) {
+      case PATH:
+        return PathDatasetBlob.class.getSimpleName();
+      case S3:
+      default:
+        return S3DatasetBlob.class.getSimpleName();
```

Instead of return S3DatasetBlob name, we should have to throw an exception instead of assuming it will have an S3 blob? cc @ravishetye @conradoverta

---

fixed